modules/message-lane/rpc/src/lib.rs

```diff
@@ -0,0 +1,218 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Module that provides RPC methods speific to message-lane pallet.
+
+use crate::error::{Error, FutureResult};
+
+use bp_message_lane::{LaneId, MessageNonce};
+use futures::{FutureExt, TryFutureExt};
+use jsonrpc_core::futures::Future as _;
+use jsonrpc_derive::rpc;
+use sc_client_api::Backend as BackendT;
+use serde::{Deserialize, Serialize};
+use sp_blockchain::{Error as BlockchainError, HeaderBackend};
+use sp_core::{storage::StorageKey, Bytes};
+use sp_runtime::{generic::BlockId, traits::Block as BlockT};
+use sp_state_machine::prove_read;
+use sp_trie::StorageProof;
+use std::sync::Arc;
+
+mod error;
+
+/// Instance id.
+pub type InstanceId = [u8; 4];
+
+/// Trie-based storage proof that the message(s) with given key(s) are sent by the bridged chain.
+pub type MessagesProof = SerializableStorageProof;
+
+/// Trie-based storage proof that the message(s) with given key(s) are received by the bridged chain.
+pub type MessagesRetrievalProof = SerializableStorageProof;
+
+/// Trie-based storage proof that the message(s) with given key(s) are processed by the bridged chain.
+pub type MessagesProcessingProof = SerializableStorageProof;
+
+/// Serializable storage proof.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct SerializableStorageProof {
+	proof: Vec<Bytes>,
+}
+
+/// Runtime adapter.
+pub trait Runtime: Send + Sync + 'static {
+	/// Return runtime storage key for given message. May return None if instance is unknown.
+	fn message_key(&self, instance: &InstanceId, lane: &LaneId, nonce: MessageNonce) -> Option<StorageKey>;
+	/// Return runtime storage key for inbound lane state. May return None if instance is unknown.
+	fn inbound_lane_data_key(&self, instance: &InstanceId, lane: &LaneId) -> Option<StorageKey>;
+}
+
+/// Provides RPC methods for interacting with message-lane pallet.
+#[rpc]
+pub trait MessageLaneApi<BlockHash> {
+	/// Returns proof-of-message(s) in given inclusive range.
+	#[rpc(name = "prove_messages")]
+	fn prove_messages(
+		&self,
+		instance: InstanceId,
```

Comment:
I do not like this argument (`instance`), but the alternative is to introduce several instances of `MessageLaneRpcHandler`, which is even worse imo. Maybe you have different opinion?

Reply:
I think it would make sense to have some kind of pallet-instances namespacing. I.e. we would have a bunch of pallets under say `eth-pow`  ID (bridge, message lanes, message dispatch, etc). Then the id would be common for these pallets.

Reply:
Ok, then I'm leaving `InstanceId` arg. Do you think it worth to change it to `String` then? Also - leaving typedef here, because we do not have any appropriate crate now (that is accessible from both node && this crate). Probably worth adding new `bp-runtime-primitives`?

Reply:
>  Do you think it worth to change it to String then?
No, I think `InstanceId` or some other couple-of-bytes typedef is fine.

> Probably worth adding new bp-runtime-primitives?
:+1: